metadata-ingestion/src/datahub/ingestion/source/mlflow.py

```diff
@@ -262,6 +281,193 @@ def _convert_run_result_type(
+            yield from hosted_dataset_reference.as_workunits()
+
+        # Add dataset references as upstreams for the run
+        if dataset_reference_urns:
+            input_edges = [
+                EdgeClass(destinationUrn=dataset_ref_urn)
+                for dataset_ref_urn in dataset_reference_urns
+            ]
+            yield MetadataChangeProposalWrapper(
+                entityUrn=str(run_urn),
+                aspect=DataProcessInstanceInputClass(inputs=[], inputEdges=input_edges),
+            ).as_workunit()
+
+    def _get_dataset_platform_from_source_type(self, source_type: str) -> Optional[str]:
+        """
+        Map MLflow source type to DataHub platform.
+
+        Priority:
+        1. User-provided mapping in config
+        2. Direct platform match in source registry
+        3. Internal mapping
+        """
+        source_type = source_type.lower()
+
+        # Check user-provided mapping
+        if (
+            self._materalize_dataset_inputs()
+            and self.config.materialize_dataset_inputs.source_mapping_to_platform
+        ):
+            platform = (
+                self.config.materialize_dataset_inputs.source_mapping_to_platform.get(
+                    source_type
+                )
+            )
+            if platform:
+                if self._is_valid_platform(platform):
+                    return platform
+                raise ValueError(
+                    f"Invalid platform '{platform}' in source_mapping_to_platform"
+                )
+
+        # Check direct platform match
+        if self._is_valid_platform(source_type):
+            return source_type
+
+        # Internal mapping
+        if source_type == "gs":
+            return "gcs"
+
+        return None
+
+    def _get_hosted_dataset_upstream(
+        self, platform: Optional[str], dataset: MLflowDataset
+    ) -> Optional[UpstreamLineageClass]:
+        """
+        Create upstream lineage for a hosted dataset.
+
+        Returns None if:
+        1. Platform is not supported
+        2. Materialization is disabled and dataset doesn't exist
+        """
+        if not platform:
+            return None
+
+        hosted_dataset_urn = DatasetUrn(platform=platform, name=dataset.name)
+
+        # When materialization is disabled, only create upstream if dataset exists
+        if (
+            self._materalize_dataset_inputs()
+            and not self.config.materialize_dataset_inputs.enabled
+        ):
+            if not self._check_dataset_existence(str(hosted_dataset_urn)):
+                return None
+
+        return UpstreamLineageClass(
+            upstreams=[UpstreamClass(str(hosted_dataset_urn), type="COPY")]
+        )
+
+    def _is_valid_platform(self, platform: str) -> bool:
+        """Check if platform is registered as a source plugin"""
+        return platform in source_registry.mapping
+
+    def _check_dataset_existence(self, dataset_urn: str) -> bool:
```

Comment:
FYI we don't check this if materialization is enabled -- we just try to upsert. 
seems like this is not guaranteed during the process bc ingestion is happening async. let me know if there's a better approach

Reply:
this exists check should not be necessary